sympy/polys/numberfields/modules.py

```diff
@@ -0,0 +1,1242 @@
+            return self.from_int_list(self.module, C, denom=d)
+        elif isinstance(other, ModuleElement):
+            a, b = self.make_compat(other)
+            if a is None:
+                return NotImplemented
+            return a * b
+        elif is_rat(other):
+            a, b = get_num_denom(other)
+            if a == b == 1:
+                return self.copy()
+            else:
+                return make_mod_elt(self.module,
+                                 self.col * a, denom=self.denom * b).reduced()
+        return NotImplemented
+
+    __rmul__ = __mul__
+
+    def _zeroth_power(self):
+        return self.module.one()
+
+    def _first_power(self):
+        return self.copy()
+
+    def __floordiv__(self, a):
+        if is_rat(a):
+            a = QQ(a)
+            if a == 0:
+                raise ZeroDivisionError
+            return self * (1/a)
+        return NotImplemented
+
+    def __rfloordiv__(self, a):
+        return a // self.over_power_basis()
+
+    def __mod__(self, a):
+        r"""
+        Reducing mod an integer *a* reduces all numerator coeffs mod $d a$, where
+        $d$ is our denominator.
+
+        For example, if we represent
+
+        $$ \frac{15 a_1 + a_0}{2} = \frac{a_1 + a_0}{2} + 7 a_1 $$
+
+        then reducing mod 7 should mean throwing away that part that is a poly
+        in the basis elements $a_i$, with content divisible by 7. But this is
+        achieved by reducing our coeffs mod $2 \times 7$.
+        """
+        if is_int(a):
+            m = a * self.denom
+            col = to_col([c % m for c in self.coeffs])
+            return type(self)(self.module, col, denom=self.denom)
+        return NotImplemented
+
+
+class PowerBasisElement(ModuleElement):
+    """
+    Subclass for ModuleElements whose module is a PowerBasis.
+    """
+
+    @property
+    def T(self):
+        return self.module.T
+
+    def numerator(self, x=None):
+        """Obtain the numerator as a polynomial over :ref:`ZZ`."""
+        x = x or self.T.gen
+        return Poly(reversed(self.coeffs), x, domain=ZZ)
+
+    def poly(self, x=None):
+        """Obtain the number as a polynomial over :ref:`QQ`."""
+        return self.numerator(x=x) // self.denom
+
+    def norm(self, T=None):
+        """Compute the norm of this number."""
+        T = T or self.T
+        x = T.gen
+        A = self.numerator(x=x)
+        return T.resultant(A) // self.denom ** self.n
+
+    def inverse(self):
+        f = self.poly()
+        f_inv = f.invert(self.T)
+        return self.module.element_from_poly(f_inv)
+
+    def __rfloordiv__(self, a):
+        return self.inverse() * a
+
+    def _negative_power(self, e, modulo=None):
+        return self.inverse() ** abs(e)
+
+
+class ModuleHomomorphism:
+    """A homomorphism from one module to another."""
+
+    def __init__(self, domain, codomain, mapping):
+        self.domain = domain
+        self.codomain = codomain
+        self.mapping = mapping
```

Comment:
The type and meaning of these parameters needs to be explained. I presume domain and codomain are both of type Module. What should `mapping` be (e.g. a dict)? A doctest example helps a lot in situations like this.

Reply:
I mapping an arbitrary callable?

Reply:
Yup, it is. Docstring added.